File: examples/java/src/main/java/org/apache/beam/examples/MinimalWordCount.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.examples;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.ServiceLoader;
import java.util.stream.StreamSupport;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.schemas.Schema;
import org.apache.beam.sdk.schemas.Schema.Field;
import org.apache.beam.sdk.schemas.Schema.FieldType;
import org.apache.beam.sdk.schemas.SchemaUtils;
import org.apache.beam.sdk.schemas.io.SchemaIOProvider;
import org.apache.beam.sdk.schemas.transforms.SchemaTransform;
import org.apache.beam.sdk.schemas.transforms.SchemaTransformProvider;
import org.apache.beam.sdk.values.PCollectionRowTuple;
import org.apache.beam.sdk.values.Row;

public class MinimalWordCount {
  public static class TransformSpec {
    public String inputId;
    public List<Object> configurationAsList = null; // Or represent as objects.
    public Row configurationAsRow = null;

    public TransformSpec(String inputId, Row configurationAsRow) {
      this.inputId = inputId;
      this.configurationAsRow = configurationAsRow;
    }

    public TransformSpec(String inputId, List<Object> configurationAsList) {
      this.inputId = inputId;
      this.configurationAsList = configurationAsList;
    }

    private SchemaTransformProvider getProvider() {
      // Check for transform provider:
      ServiceLoader<SchemaTransformProvider> provider =
          ServiceLoader.load(SchemaTransformProvider.class);

      Optional<SchemaTransformProvider> maybeProvider =
          StreamSupport.stream(provider.spliterator(), false)
              .filter(p -> p.identifier().equals(inputId))
              .findFirst();
      if (maybeProvider.isPresent()) {
        return maybeProvider.get();
      }

      // Try to look for name:read.
      int lastIndex = inputId.lastIndexOf(":");
      String inputName = inputId.substring(0, lastIndex);
      boolean isRead = inputId.substring(lastIndex + 1).equals("read");

      SchemaIOProvider ioProvider =
          StreamSupport.stream(ServiceLoader.load(SchemaIOProvider.class).spliterator(), false)
              .filter(p -> p.identifier().equals(inputName))
              .findFirst()
              .get();
      return new SchemaIOToTransformProvider(ioProvider, isRead);
    }

    public SchemaTransform getSchemaTransform() {
      SchemaTransformProvider provider = getProvider();
      return provider.from(
          configurationAsRow == null
              ? Row.withSchema(provider.configurationSchema())
                  .addValues(configurationAsList)
                  .build()
              : SchemaUtils.convertRowToSchema(configurationAsRow, provider.configurationSchema()));
    }
  }

  public static void main(String[] args) {

    PipelineOptions options = PipelineOptionsFactory.create();
    Pipeline p = Pipeline.create(options);

    boolean single = false;
    boolean avroInput = true;
    Schema schema =
        single
            ? Schema.of(Field.of("single", FieldType.STRING))
            : Schema.of(Field.of("species", FieldType.STRING), Field.of("number", FieldType.INT32));

    /*
    ArrayList<Row> records = new ArrayList<>();
    if (single) {
      records.add(Row.withSchema(schema).addValues("dog", 5).build());
      records.add(Row.withSchema(schema).addValues("cat", 9).build());
      records.add(Row.withSchema(schema).addValues("fish", 15).build());
    } else {
      records.add(Row.withSchema(schema).addValue("dog").build());
      records.add(Row.withSchema(schema).addValue("cat").build());
      records.add(Row.withSchema(schema).addValue("frog").build());
    }
    p.apply(Create.of(records).withCoder(RowCoder.of(schema))).apply(((SchemaIO) new AvroSchemaIOProvider().from("avro-input*", null, schema)).buildWriter());
    */

    FieldDropperTransformProvider provider;

    String filePrefix = "/Users/laraschmidt/Documents/beam2/files/";

    ArrayList<TransformSpec> specs = new ArrayList<>();
    if (avroInput) {
      String filename = filePrefix + (single ? "avro-single-input*" : "avro-input*");
      specs.add(new TransformSpec("avro:read", Arrays.asList(schema, filename)));
    } else {
      Row textConfig =
          Row.withSchema(Schema.builder().addStringField("location").build())
              .addValue(filePrefix + "text")
              .build();
      specs.add(new TransformSpec("text:read:v1", textConfig));
    }
    if (!single) {
      specs.add(new TransformSpec("fieldDropper", Arrays.asList("number")));
    }
    specs.add(new TransformSpec("text:write:v1", Arrays.asList("text-out")));

    PCollectionRowTuple tuple = PCollectionRowTuple.empty(p);
    Map<String, Schema> schemaMap = new HashMap<>();

    for (TransformSpec spec : specs) {
      System.out.println(spec.inputId);
      SchemaTransform transform = spec.getSchemaTransform();
      // We know we only deal with transforms with either 0 or 1 input so we know how to connect the
      // collections. To sanity check we should confirm the output collections match expected.
      if (tuple.getAll().size() == 1) {
        String input = spec.getProvider().getInputCollectionNames().get(0);
        String priorOutput = tuple.getAll().keySet().stream().findFirst().get().getId();
        tuple = PCollectionRowTuple.of(input, tuple.get(priorOutput));
        schemaMap.put(input, schemaMap.remove(priorOutput));
      }

      tuple = tuple.apply(transform.buildTransform());
      schemaMap = SchemaUtils.getSchema(transform, schemaMap);
    }

    p.run().waitUntilFinish();
  }
}

File: examples/kotlin/src/main/java/org/apache/beam/examples/kotlin/MinimalWordCount.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.examples.kotlin

import org.apache.beam.sdk.Pipeline
import org.apache.beam.sdk.io.TextIO
import org.apache.beam.sdk.options.PipelineOptionsFactory
import org.apache.beam.sdk.transforms.*
import org.apache.beam.sdk.values.KV
import org.apache.beam.sdk.values.PCollection
import org.apache.beam.sdk.values.TypeDescriptors

/**
 * An example that counts words in Shakespeare.
 *
 *
 * This class, [MinimalWordCount], is the first in a series of four successively more
 * detailed 'word count' examples. Here, for simplicity, we don't show any error-checking or
 * argument processing, and focus on construction of the pipeline, which chains together the
 * application of core transforms.
 *
 *
 * Next, see the [WordCount] pipeline, then the [DebuggingWordCount], and finally the
 * [WindowedWordCount] pipeline, for more detailed examples that introduce additional
 * concepts.
 *
 *
 * Concepts:
 *
 * <pre>
 * 1. Reading data from text files
 * 2. Specifying 'inline' transforms
 * 3. Counting items in a PCollection
 * 4. Writing data to text files
</pre> *
 *
 *
 * No arguments are required to run this pipeline. It will be executed with the DirectRunner. You
 * can see the results in the output files in your current working directory, with names like
 * "wordcounts-00001-of-00005. When running on a distributed service, you would use an appropriate
 * file service.
 */
public object MinimalWordCount {

    @JvmStatic
    fun main(args: Array<String>) {

        // Create a PipelineOptions object. This object lets us set various execution
        // options for our pipeline, such as the runner you wish to use. This example
        // will run with the DirectRunner by default, based on the class path configured
        // in its dependencies.
        val options = PipelineOptionsFactory.create()

        // In order to run your pipeline, you need to make following runner specific changes:
        //
        // CHANGE 1/3: Select a Beam runner, such as BlockingDataflowRunner
        // or FlinkRunner.
        // CHANGE 2/3: Specify runner-required options.
        // For BlockingDataflowRunner, set project and temp location as follows:
        //   val dataflowOptions : DataflowPipelineOptions = options.as(DataflowPipelineOptions::class.java)
        //   dataflowOptions.runner = BlockingDataflowRunner::class.java
        //   dataflowOptions.project = "SET_YOUR_PROJECT_ID_HERE"
        //   dataflowOptions.tempLocation = "gs://SET_YOUR_BUCKET_NAME_HERE/AND_TEMP_DIRECTORY"
        // For FlinkRunner, set the runner as follows. See {@code FlinkPipelineOptions}
        // for more details.
        //   options.as(FlinkPipelineOptions::class.java)
        //      .setRunner(FlinkRunner::class.java)

        // Create the Pipeline object with the options we defined above
        val p = Pipeline.create(options)

        // Concept #1: Apply a root transform to the pipeline; in this case, TextIO.Read to read a set
        // of input text files. TextIO.Read returns a PCollection where each element is one line from
        // the input text (a set of Shakespeare's texts).

        // This example reads a public data set consisting of the complete works of Shakespeare.
        p.apply<PCollection<String>>(TextIO.read().from("gs://apache-beam-samples/shakespeare/*"))

                // Concept #2: Apply a FlatMapElements transform the PCollection of text lines.
                // This transform splits the lines in PCollection<String>, where each element is an
                // individual word in Shakespeare's collected texts.
                .apply(
                        FlatMapElements.into(TypeDescriptors.strings())
                                .via(ProcessFunction<String, List<String>> { input -> input.split("[^\\p{L}]+").toList() })
                )
                // We use a Filter transform to avoid empty word
                .apply(Filter.by(SerializableFunction<String, Boolean> { input -> !input.isEmpty() }))
                // Concept #3: Apply the Count transform to our PCollection of individual words. The Count
                // transform returns a new PCollection of key/value pairs, where each key represents a
                // unique word in the text. The associated value is the occurrence count for that word.
                .apply(Count.perElement<String>())
                // Apply a MapElements transform that formats our PCollection of word counts into a
                // printable string, suitable for writing to an output file.
                .apply(
                        MapElements.into(TypeDescriptors.strings())
                                .via(ProcessFunction<KV<String, Long>, String> { input -> "${input.key} : ${input.value}" })
                )
                // Concept #4: Apply a write transform, TextIO.Write, at the end of the pipeline.
                // TextIO.Write writes the contents of a PCollection (in this case, our PCollection of
                // formatted strings) to a series of text files.
                //
                // By default, it will write to a set of files with names like wordcounts-00001-of-00005
                .apply(TextIO.write().to("wordcounts"))

        p.run().waitUntilFinish()
    }
}
